pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskExecutor.java

```diff
@@ -0,0 +1,247 @@
+      SegmentMetadataImpl segmentMetadata, String segmentName) {
+    SegmentGeneratorConfig config = new SegmentGeneratorConfig(tableConfig, segmentMetadata.getSchema());
+    config.setOutDir(workingDir.getPath());
+    config.setSegmentName(segmentName);
+    // Keep index creation time the same as original segment because both segments use the same raw data.
+    // This way, for REFRESH case, when new segment gets pushed to controller, we can use index creation time to
+    // identify if the new pushed segment has newer data than the existing one.
+    config.setCreationTime(String.valueOf(segmentMetadata.getIndexCreationTime()));
+
+    // The time column type info is not stored in the segment metadata.
+    // Keep segment start/end time to properly handle time column type other than EPOCH (e.g.SIMPLE_FORMAT).
+    if (segmentMetadata.getTimeInterval() != null) {
+      config.setTimeColumnName(tableConfig.getValidationConfig().getTimeColumnName());
+      config.setStartTime(Long.toString(segmentMetadata.getStartTime()));
+      config.setEndTime(Long.toString(segmentMetadata.getEndTime()));
+      config.setSegmentTimeUnit(segmentMetadata.getTimeUnit());
+    }
+    return config;
+  }
+
+  // TODO: Consider moving this method to a more appropriate class (eg ServerSegmentMetadataReader)
+  private static ImmutableRoaringBitmap getValidDocIds(String tableNameWithType, Map<String, String> configs)
+      throws URISyntaxException {
+    String segmentName = configs.get(MinionConstants.SEGMENT_NAME_KEY);
+    String server = getServer(segmentName, tableNameWithType);
+
+    // get the url for the validDocIds for the server
+    InstanceConfig instanceConfig = _clusterManagementTool.getInstanceConfig(_clusterName, server);
+    String endpoint = InstanceUtils.getServerAdminEndpoint(instanceConfig);
+    String url =
+        new URIBuilder(endpoint).setPath(String.format("/segments/%s/%s/validDocIds", tableNameWithType, segmentName))
+            .toString();
+
+    // get the validDocIds from that server
+    Response response = ClientBuilder.newClient().target(url).request().get(Response.class);
+    Preconditions.checkState(response.getStatus() == Response.Status.OK.getStatusCode(),
+        "Unable to retrieve validDocIds from %s", url);
+    byte[] snapshot = response.readEntity(byte[].class);
+    ImmutableRoaringBitmap validDocIds = new ImmutableRoaringBitmap(ByteBuffer.wrap(snapshot));
+    return validDocIds;
+  }
+
+  @VisibleForTesting
+  public static String getServer(String segmentName, String tableNameWithType) {
+    String server = null;
+    ExternalView externalView = _clusterManagementTool.getResourceExternalView(_clusterName, tableNameWithType);
+    if (externalView == null) {
+      throw new IllegalStateException("External view does not exist for table: " + tableNameWithType);
+    }
+    for (Map.Entry<String, Map<String, String>> entry : externalView.getRecord().getMapFields().entrySet()) {
```

Comment:
We can simply look up the map here. We also need to check the state of the server and pick one that is `ONLINE`
```suggestion
  Map<String, String> instanceStateMap = externalView.getStateMap(segmentName);
  if (instanceStateMap == null) {
    throw new IllegalStateException("Failed to find segment: ...");
  }
  for (Map.Entry<String, String> entry : instanceStateMap.entrySet()) {
    if (entry.getValue().equals(SegmentStateModel.ONLINE) {
      return entry.getKey();
    }
  }
  throw new IllegalStateException("Failed to find ONLINE server for segment: ...");
```